lib/src/rewrite.rs

```diff
@@ -379,34 +380,55 @@ impl<'settings, 'repo> DescendantRebaser<'settings, 'repo> {
-            // This can trigger if we abandon an empty commit, as both the empty commit and
-            // its parent are succeeded by the same commit.
-            if !new_ids.contains(id) {
-                new_ids.push(id.clone());
-            }
-        };
-        for old_id in old_ids {
-            if let Some(new_parent_ids) = self.parent_mapping.get(old_id) {
-                for new_parent_id in new_parent_ids {
-                    // The new parent may itself have been rebased earlier in the process
-                    if let Some(newer_parent_id) = self.rebased.get(new_parent_id) {
-                        add_parent(newer_parent_id);
-                    } else {
-                        add_parent(new_parent_id);
+        /// Update `new_ids`
+        fn single_substitution_round(
+            parent_mapping: &HashMap<CommitId, Vec<CommitId>>,
+            new_ids: &mut Vec<CommitId>,
+        ) -> bool {
+            let mut made_replacements = false;
+            // TODO(ilyagr): (Maybe?) optimize common case of replacements all
+            // being singletons. If CommitId-s were Copy. no allocations would be needed in
+            // that case, but it probably doesn't matter much while they are Vec<u8>-s.
+            let previously_new_ids = new_ids.split_off(0);
+            for id in previously_new_ids.into_iter() {
+                match parent_mapping.get(&id) {
+                    None => new_ids.push(id),
+                    Some(replacements) => {
+                        assert!(
+                            !replacements.is_empty(),
+                            "Found empty value for key {:?} in the parent mapping",
+                            id
+                        );
+                        made_replacements = true;
+                        new_ids.extend(replacements.iter().cloned())
                     }
-                }
-            } else if let Some(new_parent_id) = self.rebased.get(old_id) {
-                add_parent(new_parent_id);
-            } else {
-                add_parent(old_id);
-            };
+                };
+            }
+            made_replacements
+        }
+
+        let mut new_ids: Vec<CommitId> = old_ids.into();
+        let mut iterations = 0;
+        while single_substitution_round(&self.parent_mapping, &mut new_ids) {
+            iterations += 1;
+            assert!(
+                iterations <= self.parent_mapping.len(),
+                "cycle detected in the parent mapping"
+            );
+        }
+        match new_ids.as_slice() {
+            // The first two cases are an optimization for the common case of commits with <=2
+            // parents
+            [_singleton] => new_ids,
+            [a, b] if a != b => new_ids,
+            _ => {
+                // De-duplicate ids while preserving the order
+                let new_ids_set: IndexSet<CommitId> = new_ids.into_iter().collect();
+                new_ids_set.into_iter().collect()
```

Comment:
nit: `.unique()`?
https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.unique

Reply:
Thanks, I didn't know that existed!

IndexSet seems slightly superior in that it doesn't seem to clone anything, while `.unique` does. Do you still prefer the latter? I'm fine either way.

I was also thinking of using more `IndexSet`-s in general for any Set that ends up being iterated over, but I'm unsure it's worth it.

Reply:
I prefer `.unique()` since the method name describes the intent.